### PR TITLE
introduce interfaces for behavior context

### PR DESCRIPTION
**Pull request by timbussmann**
Replaces `BehaviorContext` implementations in Behaviors with interfaces in order to hide the complex internal hierarchy of the contexts and make them much easier to test.

### CODE REVIEW

**Comment by timbussmann** [comment]
@Particular/tf-handlercontext @Particular/nservicebus please review

**Comment by hmemcpy** [comment]
:+1:

File: src/NServiceBus.Core/Pipeline/Incoming/IncomingContextImpl.cs [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        protected IncomingContextImpl(string messageId, string replyToAddress, IReadOnlyDictionary<string, string> headers, BehaviorContext parentContext)
+            : base(parentContext)
+        {
+
+            MessageId = messageId;
+            ReplyToAddress = replyToAddress;
+            MessageHeaders = headers;
+        }
+
+        /// <inheritdoc />
```
Comment (SimonCropp) [inline src/NServiceBus.Core/Pipeline/Incoming/IncomingContextImpl.cs]: `<inheritdoc />` actually doenst do anything

and no need for xml doco on an internal class

**Comment by SimonCropp** [comment]
why not using the `I` prefix on interfaces and drom the `Impl` on classes?

**Comment by danielmarbach** [comment]
When tim and I talked about it we figured it is the smaller change for
users of the contexts. And it would still be aligned since we also have
things like readonlysettings without I prefix

On Thursday, December 3, 2015, Simon Cropp <notifications@github.com> wrote:

> why not using the I prefix on interfaces and drom the Impl on classes?
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/Particular/NServiceBus/pull/3148#issuecomment-161817840>
> .
>


-- 
*#NSBcon 2015 • The "All About NServiceBus" conference is back *
*• Learn more <http://go.particular.net/l/27442/2015-08-21/5d4wzp> →*

**Comment by SimonCropp** [comment]
>  it is the smaller change for users of the contexts

but most of the context have changed anyway or are new. . so are there really many "existing users"?

**Comment by timbussmann** [comment]
> but most of the context have changed anyway or are new. . so are there really many "existing users"?

valid point.
It requires less doco updates? :laughing:

File: src/NServiceBus.Core/Pipeline/Incoming/IncomingContextImpl.cs (hunk 1 above)
Comment (timbussmann) [inline src/NServiceBus.Core/Pipeline/Incoming/IncomingContextImpl.cs]: removed them :+1:

**Pull request merged**